cpp/include/kvikio/file_handle.hpp

```diff
@@ -502,45 +503,51 @@ class FileHandle {
-   * cuFileStreamRegister API, this value will not be evaluated until execution time.
-   * @param devPtr_offset Pointer to the offset relative to the bufPtr_base pointer from which to
+   * @param size_p Pointer to size in bytes to read. If the exact size is not known at the time of
+   * I/O submission, then you must set it to the maximum possible I/O size for that stream I/O.
+   * Later the actual size can be set prior to the stream I/O execution.
+   * @param file_offset_p Pointer to offset in the file from which to read. Unless otherwise set
+   * using cuFileStreamRegister API, this value will not be evaluated until execution time.
+   * @param devPtr_offset_p Pointer to the offset relative to the bufPtr_base pointer from which to
    * write. Unless otherwise set using cuFileStreamRegister API, this value will not be evaluated
    * until execution time.
-   * @param bytes_read Pointer to the bytes read from file. This pointer should be a non-NULL value
-   * and *bytes_read set to 0. The bytes_read memory should be allocated with cuMemHostAlloc/malloc/
-   * mmap or registered with cuMemHostRegister.
-   * After successful execution of the operation in the stream, the value *bytes_read will contain
-   * either:
+   * @param bytes_read_p Pointer to the bytes read from file. This pointer should be a non-NULL
+   * value and *bytes_read_p set to 0. The bytes_read_p memory should be allocated with
+   * cuMemHostAlloc/malloc/ mmap or registered with cuMemHostRegister. After successful execution of
+   * the operation in the stream, the value *bytes_read_p will contain either:
    *     - The number of bytes successfully read.
    *     - -1 on IO errors.
    *     - All other errors return a negative integer value of the CUfileOpError enum value.
    * @param stream CUDA stream in which to enqueue the operation. If NULL, make this operation
    * synchronous.
    */
-  inline void read_async(void* devPtr_base,
-                         std::size_t* size,
-                         off_t* file_offset,
-                         off_t* devPtr_offset,
-                         ssize_t* bytes_read,
-                         CUstream stream)
+  void read_async(void* devPtr_base,
+                  std::size_t* size_p,
+                  off_t* file_offset_p,
+                  off_t* devPtr_offset_p,
+                  ssize_t* bytes_read_p,
+                  CUstream stream)
   {
 #ifdef KVIKIO_CUFILE_STREAM_API_FOUND
-    CUFILE_TRY(cuFileAPI::instance().ReadAsync(
-      _handle, devPtr_base, size, file_offset, devPtr_offset, bytes_read, stream));
-    return;
-#else
-    throw CUfileException("cuFile's stream API isn't available, please build with CUDA v12.2+.");
+    if (kvikio::is_batch_and_stream_available() && !kvikio::defaults::compat_mode()) {
+      CUFILE_TRY(cuFileAPI::instance().ReadAsync(
+        _handle, devPtr_base, size_p, file_offset_p, devPtr_offset_p, bytes_read_p, stream));
+      return;
+    }
 #endif
```

Comment:
Perhaps prefer

```
#ifdef ...
  async_impl;
#else
  sync_impl;
#endif
```
To avoid the reader having to spot the early return in the true case? Or do you want the fallthrough to the sync case to happen even if the stream API is found but we're running in compat_mode?

Reply:
> Or do you want the fallthrough to the sync case to happen even if the stream API is found but we're running in compat_mode?

Yes, in compat_mode we shouldn't call cuFile